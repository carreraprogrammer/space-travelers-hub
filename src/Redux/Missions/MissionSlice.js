import  { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

export const getMissions = createAsyncThunk(
  'Missions/getMissions', // Name of the action
  async () => {
    const response = await fetch('https://api.spacexdata.com/v3/missions')
    .then((data) => data.json())

    const missionApiInfo = response.map((mission) => {
        const missions = {
          mission_id: mission.mission_id,
          mission_name: mission.mission_name,
          description: mission.description,
          reserved: false
        };
        return missions;
      }      
    )
    return missionApiInfo
      }
    )

    
const initialState = { loading: false, data: [], error: null, myMissions: [] };

const missionsSlice = createSlice(
 {
     name: 'missions',
     initialState,
     reducers: {
        joinMission(state, { payload }) {
          const newMissions = [];
          state.data.map((mission) => {
            if (mission.mission_id === payload) {
              newMissions.push({
                ...mission,
                reserved: !mission.reserved,
              });
            } else {
              newMissions.push(mission);
            }
            return newMissions;
          });
          return { ...state, data: newMissions};
        },
        filterMissions(state) {
            const joined = state.data.filter((mission) => {
                return mission.reserved === true;
            });
            return {...state, myMissions: joined }
        }
      },
     extraReducers: (Builder) => {
        Builder.addCase(getMissions.fulfilled, (state, action) => {
            state.data = action.payload;
        })
        .addCase(getMissions.pending, (state) => {
            const newState = { ...state, loading: true };
            return newState;
          })
        .addCase(getMissions.rejected, (state) => {
            const newState = { ...state, error: 'Error 404. Failed to fetch' };
            return newState;
          });
     }
 }
)

export default missionsSlice.reducer
export const { joinMission, filterMissions } = missionsSlice.actions;