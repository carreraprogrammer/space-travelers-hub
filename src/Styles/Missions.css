#table-container {
 width: 100%;
 display: flex;
 justify-content: center;
}

table {
  width: 90%;
  margin-top: 3%;
  margin-bottom: 3%;
}

th {
  padding: 10px;
  border: 1px solid gray;
  text-align: left;
}

td {
  padding: 24px 0 24px 0;
  border: 2px solid gray;
  margin: 0;
}

tbody > tr:nth-child(odd) {
  background-color: rgba(191, 191, 191, 0.514);
}

.description {
  width: 70%;
  padding-left: 5px;
  padding-right: 5px;
}

.mission {
  width: 10%;
  padding-left: 5px;
}

.statusContainer {
  width: 10%;
}

.status {
  background-color: gray;
  width: 80%;
  border-radius: 5px;
  font-size: 13px;
  text-align: center;
  margin: auto;
  color: white;
  font-weight: bold;
  padding: 2px;
}

.joinBtn {
  margin-left: 15%;
  width: 70%;
  padding: 5px;
  background-color: transparent;
  cursor: pointer;
}

.leaveBtn {
  margin-left: 15%;
  width: 70%;
  padding: 1px;
  background-color: transparent;
  cursor: pointer;
  border: 2px solid red;
  color: red;
}