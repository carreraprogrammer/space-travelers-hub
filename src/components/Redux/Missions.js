const UPDATE = 'UPDATE';
const JOIN_MISSION = 'JOIN_MISSION';

const missionsReducer = (state = [], action) => {
  switch (action.type) {
    case UPDATE:
      return action.payload;
    default:
      return state;
    case JOIN_MISSION:
      return state.map((mission) => {
        if (mission.id === action.payload) {
          return {
            ...mission,
            reserved: true,
          };
        }
        return mission;
      });
  }
};
export const showMissions = () => async (dispatch) => {
  const response = await fetch('https://api.spacexdata.com/v3/missions');
  const data = await response.json();
  dispatch({
    type: UPDATE,
    payload: data,
  });
};

export const joinMission = (id) => ({
  type: JOIN_MISSION,
  payload: id,
});

export default missionsReducer;
