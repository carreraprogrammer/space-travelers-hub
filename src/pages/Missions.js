import '../Styles/Missions.css';
import { useDispatch, useSelector } from 'react-redux';
import React, { useEffect } from 'react';
import { getMissions } from '../Redux/Missions/MissionSlice';


const Missions = () => {
  const dispatch = useDispatch();
  const state = true;

  const missions = useSelector((state) => state.Missions.data)
  useEffect(() => {
    dispatch(getMissions());
  });

  const joinBtn = () => {
   return state ? {display: 'none'} : {display: 'block'};
  }

  const leaveBtn = () => {
    return state ? {display: 'block'} : {display: 'none'};
  }

  return (
    <div id="table-container">
      <table className="missions-table">
       <thead>
         <tr>
           <th>Mission</th>
           <th>Description</th>
           <th>Status</th>
           <th></th>
         </tr>
       </thead>
       <tbody>
         {missions.map((mission) => {
          return (
          <tr key={mission.mission_id} id={mission.mission_id}>
            <td className="mission"> { mission.mission_name} </td>
            <td className="description"> { mission.description } </td>
            <td className="statusContainer"> <div className="status">NOT A MEMBER </div></td>
            <td className="joinContainer"> 
              <button className="joinBtn" style={joinBtn()}>Join Mission</button>
              <button className="leaveBtn" style={leaveBtn()}>Leave Mission</button> 
            </td>
          </tr>)
         })}
       </tbody>
      </table>
    </div>
  );
};

export default Missions;
