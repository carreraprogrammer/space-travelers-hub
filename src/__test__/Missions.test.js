import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Missions from  '../pages/Missions';
import configureStore from '../Redux/ConfigureStore';

it('Missions renders correctly', () => {
  const mission = renderer
    .create(<Provider store={configureStore}><Missions /></Provider>)
    .toJSON();
  expect(mission).toMatchSnapshot();
});