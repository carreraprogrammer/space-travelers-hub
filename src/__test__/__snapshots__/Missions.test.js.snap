// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions renders correctly 1`] = `
<div
  id="table-container"
>
  <table
    className="missions-table"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th />
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
